src/editor/Core/EditorView.re

```diff
@@ -125,11 +122,40 @@ let snapToCursorPosition = (view: t, lineHeight: int) => {
   };
 };
 
+type cursorLocation =
+  | Top
+  | Middle
+  | Bottom;
+
+let moveCursorToPosition = (~moveCursor, view, lineHeight, position) => {
+  let visibleLines = getVisibleLines(view, lineHeight);
+  /**
+     Check if there are fewer buffer lines than visibleLines, if so
+     H,M,L should only relate to the bufferlines. If the opposite is
+     true then the commands should only relate to the visible lines
+     of the buffer.
+   */
+  let lines = visibleLines > view.viewLines ? view.viewLines : visibleLines;
+
+  switch (position) {
+  | Top =>
```

Comment:
We should add helper methods to get the 'topVisibleLine' and 'bottomVisibleLine' for the current view. It looks like the current logic doesn't take that into account (ie, if you've scrolled down). Having `topVisibleLine` and `bottomVisibleLine` will make it easy to express the correct behavior for these actions when scrolled.

Reply:
I don't think I can see how based on what we have :thinking:

Reply:
I'll help out with these functions tomorrow if you're still stuck 👍 The main idea is we have the `scrollY` and the `lineHeight` available - the top line at the moment is `scrollY / lineHeight` and the bottom line is (`(scrollY + viewHeight) / lineHeight)`).

Reply:
Thanks @bryphe just needed a hint since I'm wasn't what the variables might map to, that helps a bunch. So scrolling is per line since `scrollY` is a multiplier of lineheight, which I take it means fractional amounts isn't possible not that it'd be desirable or not just checking my understanding